### PR TITLE
Implement push rule evaluation in Rust.

### PR DESCRIPTION
**Pull request by erikjohnston**
Builds on top of #13768

### CODE REVIEW

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [inline rust/src/push/evaluator.rs:146]: Note: we used to return `true` here for some bizarre reason. I moved it to return "false" as a) that makes way more sense, and b) "invalid" conditions get turned into "Condition::Unknown(_)` and should never match.

File: rust/src/push/mod.rs
Comment (erikjohnston) [inline rust/src/push/mod.rs:283]: This should probably actually be in https://github.com/matrix-org/synapse/pull/13768, though I don't think we use it in practice?

File: synapse/push/bulk_push_rule_evaluator.py
Comment (erikjohnston) [inline synapse/push/bulk_push_rule_evaluator.py:362]: This is moved from `push_rule_evaluator.py`

File: synapse/push/httppusher.py
Comment (erikjohnston) [inline synapse/push/httppusher.py:59]: This is moved from `push_rule_evaluator.py`

File: rust/src/push/evaluator.rs
Comment (MatMaul) [inline rust/src/push/evaluator.rs:87]: is the labeling of the loop purely cosmetic or something is happening here regarding lifetimes and the borrower ? Doc about those loop labels also speaks about lifetime so I am a bit confused.

File: rust/src/push/evaluator.rs
Comment (MatMaul) [inline rust/src/push/evaluator.rs:135]: The impl is separated in 2 here but I am not sure of the splitting logic.

File: rust/src/push/evaluator.rs
Comment (MatMaul) [inline rust/src/push/evaluator.rs:283]: Is a cache of compiled patterns would be useful ? Compiling regex tends to be expensive.

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:87]: The labelling is purely so we can continue of the outer loop below, via `continue 'outer`

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:135]: Ah yeah, so the first impl block is annotated with `#[pymethods]` which mean that those methods are exposed to the python layer, and as such have strict requirements on what types can be used. The second impl block is just standard Rust methods.

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:283]: Oh, good point. We did have a cache before too. Let me quickly run a benchmark.

File: rust/src/push/evaluator.rs
Comment (MatMaul) [reply rust/src/push/evaluator.rs:87]: Right I managed to miss the fact that they were 2 loops here hum.

File: rust/src/push/evaluator.rs
Comment (MatMaul) [inline rust/src/push/evaluator.rs:118]: It seems like `matches` is unused globally or I am missing something ?

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:283]: Interesting numbers come out of it (annoying I can't commit the benchmark code, as benchmarking doesn't work in a `cdylib` crate):

```
test bench_glob_whole   ... bench:      28,814 ns/iter (+/- 1,209)
test bench_glob_word    ... bench:     409,264 ns/iter (+/- 47,114)
```

So compiling word based regex can take almost half a millisecond on my laptop. If we disable unicode awareness when compiling the regex they end up taking about the same amount of time.

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:118]: It's used in the python unit tests to test matching against a condition

**Comment by erikjohnston** [comment]
(I need to do something about the glob-to-regex cache)

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs]: Instead of caching the regex, I've implemented a "fast path", which allows us to avoid compiling the regex unless necessary.

**Comment by erikjohnston** [comment]
A local test of this on postgres takes the time to send a message in a room with 5k local users from ~3s to ~500ms, with the actual calculating of push actions a tiny fraction of that.

File: rust/src/push/evaluator.rs
Comment (clokep) [reply rust/src/push/evaluator.rs:163]: I always found this really weird, I wonder if we should land that now in the current code (if we think this is going to take a bit of review cycles before landing) to ensure it doesn't break clients?

Comment (erikjohnston) [reply rust/src/push/evaluator.rs:163]: To be clear: we were returning true under this condition in the Python code, so this isn't a regression or anything

Comment (clokep) [reply rust/src/push/evaluator.rs:163]: But now we're returning `false`? Maybe Im' misreading the code...

**Comment by erikjohnston** [comment]
Note, the files in `rust/benches/` are benchmarks, which can be run with `cargo bench` (though you need nightly).

I'd recommend reviewing this as chunks:

1. `rust/src/push/utils.rs` — contains some helper functions that mimic those in Python
2. `rust/src/push/evaluator.rs`, `stubs/` — the actual evaluator code
3. The changes to `synapse/` — where we use the new evaluator and delete the old one
4. The benchmarks.

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:163]: Sorry for not being clear. This is the code currently on develop and in releases:

https://github.com/matrix-org/synapse/blob/87fe9db4675e510ea9c0234429b4773341c4e86d/synapse/push/push_rule_evaluator.py#L211-L216

File: rust/Cargo.toml
Comment (clokep) [inline rust/Cargo.toml:15]:
```suggestion
# We generate a `cdylib` for Python and a standard `lib` for running
# tests/benchmarks.
```

File: rust/src/push/utils.rs
Comment (reivilibre) [inline rust/src/push/utils.rs:25]: docstring a little bit inaccurate; to my reading this matches the non-wildcards as group 1 followed by the wildcards as group 2.
Does it concern us that this regex can match an empty string? I suspect not if you've tried this

File: rust/src/push/evaluator.rs
Comment (reivilibre) [inline rust/src/push/evaluator.rs:32]: doc comment please :-)

File: rust/src/push/evaluator.rs
Comment (reivilibre) [inline rust/src/push/evaluator.rs:147]: surprised that an error will be considered to match. Do we have some reasoning that we could also write in the comments?

File: rust/src/push/evaluator.rs
Comment (reivilibre) [inline rust/src/push/evaluator.rs:330]:
```suggestion
    /// Match the member count against an 'is' condition
    /// The `is` condition can be things like '>2', '==3' or even just '4'.
    fn match_member_count(&self, is: &str) -> Result<bool, Error> {
```

File: rust/src/push/evaluator.rs
Comment (reivilibre) [reply rust/src/push/evaluator.rs:32]: If I'm reading correctly, this would accept `=====>>>>><<<<42`.
possibly better to just list out the cases as alternate choices `(==|<=|>=|>|<|...)`?

Comment (reivilibre) [reply rust/src/push/evaluator.rs:32]: This should say `[0-9]+` or else I can just supply `>`...

File: rust/src/push/evaluator.rs
Comment (reivilibre) [inline rust/src/push/evaluator.rs:332]: not sure where the `map_or` default would actually be used, if ever here?

File: rust/src/push/evaluator.rs
Comment (reivilibre) [inline rust/src/push/evaluator.rs:355]: I expect we'll be doing this kind of thing a lot in tests, is it worth importing some map macro (perhaps as a dev/test dependency) to make it more natural?

File: synapse/push/bulk_push_rule_evaluator.py
Comment (reivilibre) [inline synapse/push/bulk_push_rule_evaluator.py:341]: why not:
```suggestion
            if "notify" in actions:
```
am I missing something?

**Review by reivilibre — CHANGES_REQUESTED** [review]
This looks reasonable to me

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:147]: Oh, woops. This is a holdover from when we considered unknown conditions to match. This has changed. Will update.

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:332]: The regex matches plain numbers, like `3`, which are an exact match.

File: synapse/push/bulk_push_rule_evaluator.py
Comment (erikjohnston) [reply synapse/push/bulk_push_rule_evaluator.py:341]: Ah, I think this used to be `Option<Vec<Action>>`, will update.

File: rust/src/push/utils.rs
Comment (erikjohnston) [reply rust/src/push/utils.rs:25]: Matching the empty string is fine, it'll cause us to compile a regex of an empty string (which is what we want). Plus in practice the usage of this is always gated behind checking if the glob has any wildcards in.

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs]: So this is a straight up copy of the behaviour of the existing rules, we could make the regex stricter but it doesn't actually matter.

File: rust/src/push/evaluator.rs
Comment (reivilibre) [reply rust/src/push/evaluator.rs:338]: I mean that it looks like the capture group 1 should always exist, so defaulting to `==` seems superfluous.
e.g. As far as I can tell, `captures[1]` (maybe with a &-borrow or .as_str() if necessary) does the same thing

**Review by reivilibre — APPROVED** [review]
LGTM (though not sure if `map_or` is really necessary, but up to you)

File: rust/src/push/evaluator.rs
Comment (erikjohnston) [reply rust/src/push/evaluator.rs:338]: Oh, hmm, yes because it will get the empty string

Comment (erikjohnston) [reply rust/src/push/evaluator.rs:338]: We still need to unwrap the `Option` so I am somewhat minded to leave it as is?

Comment (reivilibre) [reply rust/src/push/evaluator.rs:338]: presumably `captures[1]` wouldn't be an option, but yes whatever works

**Comment by erikjohnston** [comment]
Woo, thanks all :tada:

**Pull request merged**
